cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java

```diff
@@ -259,25 +286,94 @@ public void visitEnd() {
-                                                                 AuthEnforce.class.getSimpleName()));
+                                "with name '%s' in class '%s' for method '%s' whereas " +
+                                "it was specified in '%s' annotation", name,
+                        className, methodName,
+                        AuthEnforce.class.getSimpleName()));
               }
             }
             entityPartDetails.add(entityPart);
           }
+          // verify that the entity parts specified in annotation is valid to do enforcement on the given enforceOn
+          verifyEntityParts(entityPartDetails, nodeProcessor.getEnforceOn(), parameterDetails, paraPositions);
           // Store all the information properly for the second pass
           methodAnnotations.put(new Method(methodName, methodDesc),
                                 new AnnotationDetail(entityPartDetails, nodeProcessor.getEnforceOn(),
-                                                     nodeProcessor.getActions(), paramAnnotation));
+                                                     nodeProcessor.getActions(), paraPositions));
+        }
+
+        /** Helper Methods **/
+
+        private void verifyEntityParts(List<EntityPartDetail> entityPartDetails, Type enforceOn,
+                                       Map<Integer, ParameterDetail> parameterAnnotationNode,
+                                       Map<String, Integer> paramAnnotation) {
+          Type entityPartType = getEntityPartType(parameterAnnotationNode, paramAnnotation, entityPartDetails.get(0));
+          // TODO: Later change this to even work with parent type as we will support enforceOn parent type
+          // if the first entity part is not string then it should be on same type specified in enforce on
+          if (!entityPartType.equals(Type.getType(String.class))) {
+            Preconditions.checkArgument(entityPartType.equals(enforceOn), "Found invalid entity type '%s' for " +
+                                                "enforceOn '%s' in annotation on '%s' method in '%s' class.",
+                    entityPartType.getClassName(), enforceOn.getClassName(), methodName, className);
+            Preconditions.checkArgument(getRequiredSize(enforceOn) != 0, "Invalid enforceOn %s provided.",
+                    enforceOn.getClassName());
+          } else {
+            // Since the entity part/parts provided is of String type so validate that we have sufficient part for
+            // EntityId creation
+            // verify that all parts are of type string
+            for (EntityPartDetail entityPartDetail : entityPartDetails) {
+              entityPartType = getEntityPartType(parameterAnnotationNode, paramAnnotation, entityPartDetail);
+              Preconditions.checkArgument(entityPartType.equals(Type.getType(String.class)),
+                      "Found part %s of type %s in a multiple part entity " +
```

Comment:
My Intellij is messed up :( Even after importing settings not working well. Point out some example so that I can fix it.